src/interface/swarm.cpp

```diff
@@ -317,28 +354,469 @@ void Swarm::Defrag() {
     pmb->par_for(
-        "Swarm::DefragReal", 0, max_active_index_, KOKKOS_LAMBDA(const int n) {
-          if (from_to_indices(n) >= 0) {
-            vec(from_to_indices(n)) = vec(n);
+        "Unpack buffers", 0, total_received_particles - 1, KOKKOS_LAMBDA(const int n) {
+          int sid = new_indices(n);
+          int nid = neighbor_index(n);
+          int bid = buffer_index(n);
+          for (int i = 0; i < real_vars_size; i++) {
+            vreal(i, sid) = bdvar.recv[nid](bid * particle_size + i);
+          }
+          for (int i = 0; i < int_vars_size; i++) {
+            vint(i, sid) = static_cast<int>(
+                bdvar.recv[nid]((real_vars_size + bid) * particle_size + i));
+          }
+
+          double &x = vreal(ix, sid);
+          double &y = vreal(iy, sid);
+          double &z = vreal(iz, sid);
+          // TODO(BRR) Don't hardcode periodic boundary conditions
+          if (x < swarm_d.x_min_global_) {
+            x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);
+          }
+          if (x > swarm_d.x_max_global_) {
+            x = swarm_d.x_min_global_ + (x - swarm_d.x_max_global_);
+          }
+          if (y < swarm_d.y_min_global_) {
+            y = swarm_d.y_max_global_ - (swarm_d.y_min_global_ - y);
+          }
+          if (y > swarm_d.y_max_global_) {
+            y = swarm_d.y_min_global_ + (y - swarm_d.y_max_global_);
+          }
+          if (z < swarm_d.z_min_global_) {
+            z = swarm_d.z_max_global_ - (swarm_d.z_min_global_ - z);
+          }
+          if (z > swarm_d.z_max_global_) {
+            z = swarm_d.z_min_global_ + (z - swarm_d.z_max_global_);
           }
         });
   }
 
-  // Update max_active_index_
-  max_active_index_ = num_active_ - 1;
+  bool all_boundaries_received = true;
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+    if (bdvar.flag[nb.bufid] == BoundaryStatus::arrived) {
+      bdvar.flag[nb.bufid] = BoundaryStatus::completed;
+    } else if (bdvar.flag[nb.bufid] == BoundaryStatus::waiting) {
+      all_boundaries_received = false;
+    }
+  }
+
+  if (all_boundaries_received) {
+    return true;
+  } else {
+    return false;
+  }
+}
+void Swarm::PackAllVariables(SwarmVariablePack<Real> &vreal,
+                             SwarmVariablePack<int> &vint) {
+  PackIndexMap rmap, imap;
+  std::vector<std::string> real_vars;
+  std::vector<std::string> int_vars;
+  for (auto &realVar : realVector_) {
+    real_vars.push_back(realVar->label());
+  }
+  int real_vars_size = realVector_.size();
+  int int_vars_size = intVector_.size();
+  for (auto &intVar : intVector_) {
+    int_vars.push_back(intVar->label());
+  }
+  vreal = PackVariablesReal(real_vars, rmap);
+  vint = PackVariablesInt(int_vars, imap);
+}
+
+void Swarm::PackAllVariables(SwarmVariablePack<Real> &vreal, SwarmVariablePack<int> &vint,
+                             PackIndexMap &rmap, PackIndexMap &imap) {
```

Comment:
The logic for this method and the one above is identical except for the interface consider putting it into a sub method.

Reply:
Removed this method and now there are just separate `PackAllVariablesInt` and `PackAllVariablesReal` calls. These have similar logic but are public functions that wrap the private variables `intVector_` and `realVector_` (which also can't just be templated on).